Add support for in-band console diagnostics

AloisMahdal opened the pull request:
Based on @vinzenz's earlier pull request #527.


Specification
=============

When upgrade fails in initramfs, log files and other assets will be tarballed and, using a special encoding method based on Base64, dumped to the console.

The encoding is designed to be resillient towards various forms of corruption such as other things writing to the console (mid-line) or line breaking (often seen on s390x machines in our Beaker pool).

By default, the dump functionality is disabled, but can be enabled by adding kernel option `rd.upgrade.inband`:

 *  If kernel option value is `onerror`, dump is done only if upgrade (systemd-nspawn) finished with non-zero exit status.

 *  If kernel option value is `always`, dump is done regardless of the exit status.

 *  If kernel option value is `never`, dump is not done, regardless of the exit status.

 *  If kernel option is not present, set to empty string or an unknown value, the dump is not done.  (Same as `never`.)

In order to retrieve the dumped tarball, one has to collect console log (UTF-8 is assumed) and run decoder (Python 3) included in the utils/ subdirectory:

    utils/ibdmp-decode console.log ibdmp.tar.xz

This will result in ibdmp.tar.xz, containing the set of files encoded by the upgrade dracut module.


Implementation notes
====================

In the dracut module, the `collect_and_dump_debug_data()` function is where the tarball is created.  This function passes the tarball to `ibdmp()`, which takes care of the encoding and dumping.

To make sure the data is recoverable even if the console log is affected by irrelevant writes or later corruption, the `ibdmp()` function employs lot of redundancy:  after splitting the base64-encoded tarball to small chunks and pre-pending it with some meta-data, it will perform 3 dumps, each separated with progressively increasing delay.

That is, each dump consists of one "dump header" and a large number of small (40 characters by default) "chunks".

The decoder will first find all meta-data headers (learning how many chunks to look for), then collect all chunks from all dumps and merge them according to prevalency, ie. if same chunk appears twice with value "foo" but once with value "bar", the second value is disregarded.

Finally all the chunks are joinded and decoded, resulting in original tarball file, which is finally checked against MD5 sum included in the dump header.

The number of dumps and chunk size can be controlled by integers `IBDMP_ITER` (number of dumps, 3 by default) and `IBDMB_CHUNKSIZE` length of each chunk in characters, 40 by default).  These parameters are purely internal constants; there's no intent to let user control them.

pirat89 commented:
be aware of trailing spaces :)

examon commented:
leapp-ci build

AloisMahdal commented:
s390x, dumping 3 times, wrapping at 40 chars:

fulll console.log: 3.0M
 * size without in-band-debug: 962K
 * in-band-debug: 865K
 * spam caused by in-band debug function running in `set -x` mode: 1.3M

@vinzenz I think it's bearable, and if we neutralize the `set -x` spam; we would "just double" the console.log.

AloisMahdal commented:
The good news is that here's the decoder (*warning:* outputs binary! call with redirect, eg. `./inband_dec.py console.log > console.tar.xz`):

    #!/usr/bin/python3

    import base64
    import collections
    import hashlib
    import re
    import sys

    RE_HEADER = r'_leapp_debug:[0-9]+/[0-9]+[|]chunks=[0-9]+,md5=[0-9a-f]+[|]'
    RE_CHUNK = r'_leapp_debug:[0-9]+/[0-9]+[|][0-9]+:[A-Za-z0-9+/=]+[|]'


    class Chunk:

        @classmethod
        def from_raw(cls, raw):
            assert raw.startswith('_leapp_debug:')
            areas = raw.split('|')
            parts = areas[1].split(':')
            return cls(
                cord=int(parts[0]),
                cdata=str(parts[1]),
            )

        def __init__(self, cord, cdata):
            self.cord = cord
            self.cdata = cdata

        def __hash__(self):
            return hash((self.cord, self.cdata))


    class Header:

        @classmethod
        def from_raw(cls, raw):
            parts = raw.split('|')
            _, stats = parts[0].split(':')
            pairs = parts[1].split(',')
            assert pairs[0].startswith('chunks=')
            assert pairs[1].startswith('md5=')
            return cls(
                chunks=int(pairs[0].split('=')[1]),
                md5=str(pairs[1].split('=')[1]),
                csets=int(stats.split('/')[1]),
            )

        def __init__(self, chunks, md5, csets):
            self.chunks = chunks
            self.md5 = md5
            self.csets = csets

        def __eq__(self, othr):
            return (self.chunks, self.md5) == (othr.chunks, othr.md5)

        def __hash__(self):
            return hash((self.chunks, self.md5))


    class ChunkCounter:

        def __init__(self, header, chunks_raw):
            self.header = header
            self._chunks = [None] * header.chunks
    #       sys.stderr.write('header.chunks=%r\n' % header.chunks)
            for cr in chunks_raw:
                c = Chunk.from_raw(cr)
    #           sys.stderr.write('c.cord=%r\n' % c.cord)
                if self._chunks[c.cord-1] is None:
                    self._chunks[c.cord-1] = collections.Counter([c])
                else:
                    self._chunks[c.cord-1].update([c])

        @property
        def chunks(self):
            out = []
            i = 0
            for c in self._chunks:
                i += 1
                if c is None:
                    sys.stderr.write('Missing chunk id: %d/%d\n'
                                     % (i, self.header.chunk))
                out.append(c.most_common()[0][0].cdata)
            return out

        def dump64(self):
            for c in self.chunks:
                print(c)

        def decode(self):
            tarball = base64.b64decode(''.join(self.chunks))
            tar_md5 = hashlib.md5(tarball).hexdigest()
            assert tar_md5 == self.header.md5
            sys.stdout.buffer.write(tarball)


    if __name__ == '__main__':
        headers_raw = []
        chunks_raw = []
        with open(sys.argv[1]) as f:
            for line in f.readlines():
                for m in re.findall(RE_HEADER, line):
                    headers_raw.append(m)
                for m in re.findall(RE_CHUNK, line):
                    chunks_raw.append(m)

        headers_cnt = collections.Counter([
            Header.from_raw(h) for h in headers_raw
        ])
        header = headers_cnt.most_common()[0][0]
        ccounter = ChunkCounter(header, chunks_raw)
    #   ccounter.dump64()
        ccounter.decode()
    #   print(chunks_raw)

the bad news is that I've decoded my data and turns out that it's only journalctl.log.

Ie. the the above computation **only counts journtalctl.log**.  for some reason my collector does not pick up any code.

vinzenz commented:
Tested on my machine works perfectly fine. Follow up needed to make this optional however.

AloisMahdal commented:
Where should I place the decoder?  cc @Rezney

pirat89 commented:
@AloisMahdal e.g. into the utils or into the snippets repository seems ok to me.

AloisMahdal commented:
Improvements since the last version:

 *  When searching for chunks it works with window of 2 lines so line breaks can be fixed.
 *  Added some meaningful debugging and error messaging and exit status policy.
 *  Couple of fixes.
 *  Added Python docstrings & Bash docstrings.

What still needs to be done:

 *  RTD docs (part of the original spec in our Jira); the docstrings are probably not suitable.
 *  Make it optional as @vinzenz suggested above.

AloisMahdal commented:
@vinzenz I'd like to make the dumper optional.  I suggest 3 modes affecting whether or not `collect_and_dump_debug_data()` is called:
 * off - never,
 * on error - ie. on leapp error,
 * always - always (OAMG-QE would use this for a time to help debug long-standing issues such as rhel7 kernel boot).

I'm not sure what's the proper way of controlling options in the initramfs (kernel options? via leapp?) and how to access them in the dracut code.

WDYT?  Cc @shaded-enmity, @Rezney, @pirat89, anyone else?

Rezney commented:
@AloisMahdal thanks for working on this. I like the 3 mode idea. For controlling, a kernel cmdline option should be ok. I will leave the proper access in dracut question for @vinzenz

AloisMahdal commented:
I've added the kernel option logic (separate commit for now, i will squash it prior to review):

 *  if `rd.upgrade.inband=never` - dump is never produced, regardless of leapp exit status.
 *  if `rd.upgrade.inband=always` - dump is always produced, regardless of leapp exit status.
 *  if `rd.upgrade.inband=onerror` - dump is produced, if exit status was other than zero.
 *  without the option at all, we act as if the value was `never`,
 *  if the option was provided but with unknown value, warning is printed and we act as if the value was `never`.

@vinzenz @Rezney, others, WDYT?

shaded-enmity commented on repos/system_upgrade/el7toel8/actors/commonleappdracutmodules/files/dracut/85sys-upgrade-redhat/do-upgrade.sh:
```diff
@@ -17,6 +17,149 @@ NSPAWN_OPTS="--capability=all --bind=/sys --bind=/dev --bind=/dev/pts --bind=/ru
+# first emission is done immediately, second after 10s, and the
+# third one after 20s.
+#
+IBDMP_ITER=3
+
+#
+# Size of one chunk payload
+#
+IBDMB_CHUNKSIZE=40
```
The whole archive will be dumped to the console `IBDMP_ITER` times? 

As for `IBDMB_CHUNKSIZE`:
1. Shouldn't it be `IBDMP_CHUNKSIZE`?
2. It's size in bytes / characters?

vinzenz replied:
> The whole archive will be dumped to the console `IBDMP_ITER` times?
> 
> As for `IBDMB_CHUNKSIZE`:
> 
> 1. Shouldn't it be `IBDMP_CHUNKSIZE`?
Good catch
> 2. It's size in bytes / characters?
Since the dump is supposed to be BASE64 and the data is supposed to be UTF-8, the set of characters would cause size in bytes == number of characters

pirat89 replied:
@AloisMahdal you should probably check the script by shellcheck

AloisMahdal replied:
yeah, it's number of b64-enoced chars per chunk, which by definition will be ascii, ie. bytes.

I will clarify it in the docstring.

FWIW, this is not directly about the console size footprint (although it does affect how much meta-data there will be) but more about "how big a string can i safely print without something messing with it".  (Ie. the first value is roughly half of the "standard" 80-wide terminal.)

AloisMahdal replied:
> @AloisMahdal you should probably check the script by shellcheck

I have (and the code *I*'m adding is shellcheck-ok), but the typo would not be caught by ShellCheck (it was everywhere).

AloisMahdal replied:
Anyway, I've fixed the typo 827e66e1 and clarified the docstring 5b4c85c

shaded-enmity commented on repos/system_upgrade/el7toel8/actors/commonleappdracutmodules/files/dracut/85sys-upgrade-redhat/do-upgrade.sh:
```diff
@@ -17,6 +17,149 @@ NSPAWN_OPTS="--capability=all --bind=/sys --bind=/dev --bind=/dev/pts --bind=/ru
+    cp -vr "$NEWROOT/var/log/leapp" \
+          "$data/var/log"
+    tar -cJf "$tmp/data.tar.xz" "$data"
+    ibdmp "$tmp/data.tar.xz"
```
Hmm, I'm thinking of two possible modes here:

1. We keep the data file in well known location (`rd.inband=keep`?)
2. We remove the data after they have been dumped (think systems where `/tmp` is not `tmpfs` and we're potentially leaking data to `/tmp`)

AloisMahdal replied:
> Hmm, I'm thinking of two possible modes here:
> 
> 1. We keep the data file in well known location (`rd.inband=keep`?)

I'd keep it as RFE for next PR.

> 2. We remove the data after they have been dumped (think systems where `/tmp` is not `tmpfs` and we're potentially leaking data to `/tmp`)

Good point, seems relevant to our overall function of the initramfs phase.

Ie. if leaking data matters, I imagine we should have general policy that applies not just to `collect_and_dump...*` but to all of our dracut code.  (The `keep` thing should be treated in the same context.)

AloisMahdal replied:
After some pondering, I'm simply adding

    rm -r "$tmp"

after the dump. 9543e99

That should solve the potential privacy risk of leaving the dump around.

shaded-enmity commented on repos/system_upgrade/el7toel8/actors/commonleappdracutmodules/files/dracut/85sys-upgrade-redhat/do-upgrade.sh:
```diff
@@ -17,6 +17,149 @@ NSPAWN_OPTS="--capability=all --bind=/sys --bind=/dev --bind=/dev/pts --bind=/ru
+    md5=$(md5sum "$tarball" | sed 's/ .*//')
+    chunks=$(wc -l <"$tmp/b64")
+    (
+        set +x
```
This doesn't need to be here, right?

AloisMahdal replied:
I think it does.  --  Note that `+x` *disables* the shell debug mode.

So if I drop it, the shell will re-print every upcoming arithmetic and `echo`, making the log **massive**.

shaded-enmity replied:
oh right, the perks of shell using `+` to disable things and `-` to disable things :D

AloisMahdal replied:
Yeah, no blame though, POSIX might be older than modern use of `+` and `-` in mathematics :D

shaded-enmity commented on repos/system_upgrade/el7toel8/actors/commonleappdracutmodules/files/dracut/85sys-upgrade-redhat/module-setup.sh:
```diff
@@ -73,6 +76,13 @@ install() {
+    inst_binary base64
+    inst_binary tar
+    inst_binary xz
+    inst_binary sha256sum
```
I don't see `sha256sum` being used anywhere in the code, is it a dependency of a dependency?

AloisMahdal replied:
Good point, removing.

shaded-enmity commented on utils/ibdmp-decode:
```diff
@@ -0,0 +1,261 @@
+        self.payload = payload
+
+    def __hash__(self):
+        return hash((self.ordinal, self.payload))
```
Why is the `ordinal` part of the hash? Isn't the `payload` the identity element (content-addressable)?

AloisMahdal replied:
Not if it's corrupted.

(That is, if i remember my thinking -- from 5 months ago -- correctly.)

AloisMahdal replied:
Umm, you are probably right.

shaded-enmity replied:
Hmm, is this part even needed? I don't see `_Chunk` being hashed anywhere.

AloisMahdal replied:
I think the hashability is needed because the use of `collections.Counter` relies on it.

Ie. check out `ChunkCounter.__init__` -- I'm loading the chunks into a huge dictionary (`ChunkCounter._bagset`) where keys are chunk ordinals, and each value is `collections.Counter` -- that way I can load chunks in any order and `collections.Counter` will do the maths.

You're probably right that the `ordinal` does not need to be part of the hash just for this to work.  OTOH, I would not say the payload is content-addressable: one could have "same" chunk at different positions and semantically it would not really be "the same chunk" -- so even if it's not functionally necessary (`_Chunk` is internal, after all) it seems more correct to include it.

WDYY?

shaded-enmity replied:
So on line 173 I see:
```
self._bagset[c.ordinal].update([c.payload])
```

So it seems that `c` doesn't need to be hashable because it's members are put into the structure and not `c` itself?

shaded-enmity replied:
nvm it probably gets hashed at line 130

AloisMahdal replied:
L130 is hashing `Header.chunks` but that's just int

AloisMahdal replied:
IOW, you're right, neither `Header` nor `_Chunk` actually need `__hash__` -- it's likely just remainder from some earlier attempt ising `set` or something.

shaded-enmity commented on utils/ibdmp-decode:
```diff
@@ -0,0 +1,261 @@
+        self.md5 = md5
+        self.csets = csets
+
+    def __eq__(self, othr):
```
Make sure to also provide `__neq__` to save some headaches later on (right, @pirat89 ? :))

AloisMahdal replied:
I was under the impression that one does not need to do that in Python 3 anymore... but I might be wrong.  I've added `__neq__` in abf4c26

shaded-enmity commented on utils/ibdmp-decode:
```diff
@@ -0,0 +1,261 @@
+        _, stats = parts[0].split(':')
+        pairs = parts[1].split(',')
+        try:
+            assert pairs[0].startswith('chunks=')
```
Don't `assert` → `except AssertionError` in non-test code. Assertions are skipped when Python is running with `-O` or `-OO` so they shouldn't be relied on for control flow (pretty much as all exception handling).

AloisMahdal replied:
OK, 18c9380 replaces asserts (here but also in `_Chunk.from_raw1` around L50) with `if` and uses "classic" `LOG_WARN` + `IbdmbDecodeError` combo.

And yes, it's `IbdmbDecodeError`, and yes, it's `b`, and yes, it's the same f-ing typo again -- see, @pirat89 ? -- even pylint does not help :D  fixed in 430badd

AloisMahdal commented:
I found a minor bug in the decoder chunk loading code: in case N chunks are completely lost (in all dumps), when loading chunks, the decoder would act as if there's N less overall chunks to be loaded (it would look for indices 1 .. total-N, instead of 1..total).  Ie. the number of expected chunks would be lowered by the fact that when raw-loading, some were never seen.

This would affect only cases where the data is badly corrupted anyway (in which case the MD5 check would fail), but it would make debug log misleading, so I went ahead and fixed it in 531f3c2;  now the loader will always use the number from the header.

AloisMahdal commented:
I found a minor bug in the decoder chunk loading code: in case N chunks are completely lost (in all dumps), when loading chunks, the decoder would act as if there's N less overall chunks to be loaded (it would look for indices 1 .. total-N, instead of 1..total).  Ie. the number of expected chunks would be lowered by the fact that when raw-loading, some were never seen.

This would affect only cases where the data is badly corrupted anyway (in which case the MD5 check would fail), but it would make debug log misleading, so I went ahead and fixed it in 531f3c2;  now the loader will always use the number from the header.

AloisMahdal commented:
...aand squashed commits from review back to one.

AloisMahdal commented:
....aand rebased on `master`.

shaded-enmity reviewed (commented):
Code-wise LGTM, still need to try it out. :+1:

pirat89 commented:
leapp-ci build

pirat89 commented:
leapp-ci build

shaded-enmity reviewed (approved):
LGTM. One thing that I just realized is documentation, but let's not block on it - you did a good job describing the interface in the commit message and the docs live in different repository anyway. Can you create a task to update https://github.com/oamg/leapp/tree/master/docs/source ?
